spec/spec.md

```diff
@@ -0,0 +1,2604 @@
+   mul(e1, e2)                        w~1~ + w~2~                           p~1~ + p~2~
+  ------------- ------------------------------------------------------- -------------------
+
+  ------------ ----------------------------- ----------
+                                             
+                                             
+   pad(e, n)           max(w~e~, n)             p~e~
+   shl(e, n)             w~e~ + n               p~e~
+   shr(e, n)    max(w~e~ - n, max(1, p~e~))     p~e~
+   incp(e, n)            w~e~ + n             p~e~ + n
+   decp(e, n)            w~e~ - n             p~e~ - n
+   setp(e, n)         w~e~ - p~e~ + n            n
+  ------------ ----------------------------- ----------
+
+  -------------- ----------------------- ------
+                                         
+                                         
+   dshl(e1, e1)   w~e1~ + 2`^`w~e2~ - 1   p~e~
+   dshr(e1, e2)           w~e1~           p~e~
+  -------------- ----------------------- ------
+
+# Namespaces
+
+All modules in a circuit exist in the same module namespace, and thus
+must all have a unique name.
+
+Each module has an identifier namespace containing the names of all port
+and circuit component declarations. Thus, all declarations within a
+module must have unique names. Furthermore, the set of component
+declarations within a module must be *prefix unique*. Please see section
+[11.2](#prefix_unique)
+for the definition of prefix uniqueness.
+
+Within a bundle type declaration, all field names must be unique.
+
+Within a memory declaration, all port names must be unique.
+
+During the lowering transformation, all circuit component declarations
+with aggregate types are rewritten as a group of component declarations,
+each with a ground type. The name expansion algorithm in section
+[11.1](#expansion_algorithm){reference-type="ref"
+reference="expansion_algorithm"} calculates the names of all replacement
+components derived from the original aggregate-typed component.
+
+After the lowering transformation, the names of the lowered circuit
+components are guaranteed by the name expansion algorithm and thus can
+be reliably referenced by users to pair meta-data or other annotations
+with named circuit components.
+
+## Name Expansion Algorithm {#expansion_algorithm}
+
+Given a component with a ground type, the name of the component is
+returned.
+
+Given a component with a vector type, the suffix `$`*i* is appended to
+the expanded names of each sub-element, where *i* is the index of each
+sub-element.
+
+Given a component with a bundle type, the suffix `$`*f* is appended to
+the expanded names of each sub-element, where *f* is the field name of
+each sub-element.
+
+## Prefix Uniqueness {#prefix_unique}
+
+The *symbol sequence* of a name is the ordered list of strings that
+results from splitting the name at each occurrence of the '\$'
+character.
+
+A symbol sequence $a$ is a *prefix* of another symbol sequence $b$ if
```

Comment:
This doesn't render.

Reply:
Fixed in the rebase:

<img width="835" alt="Screen Shot 2022-02-16 at 01 18 58" src="https://user-images.githubusercontent.com/1018530/154207656-8e56b696-3a92-469f-814f-8d44227d2886.png">